x-pack/plugins/spaces/server/lib/saved_objects_client/spaces_saved_objects_client.js

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { DEFAULT_SPACE_ID } from '../../../common/constants';
+import { isTypeSpaceAware } from './lib/is_type_space_aware';
+import { getSpacesQueryParams } from './lib/query_params';
+
+export class SpacesSavedObjectsClient {
+  constructor(options) {
+    const {
+      request,
+      baseClient,
+      spacesService,
+      types,
+    } = options;
+
+    this.errors = baseClient.errors;
+
+    this._client = baseClient;
+    this._request = request;
+    this._types = types;
+
+    this._spaceUrlContext = spacesService.getUrlContext(this._request, '');
+  }
+
+  async create(type, attributes = {}, options = {}) {
+
+    const spaceId = await this._getSpaceId();
+    const shouldAssignSpaceId = spaceId !== DEFAULT_SPACE_ID && isTypeSpaceAware(type);
+
+    if (shouldAssignSpaceId) {
+      options.extraBodyProperties = {
+        ...options.extraBodyProperties,
+        spaceId: await this._getSpaceId()
+      };
+    }
+
+    return await this._client.create(type, attributes, options);
+  }
+
+  async bulkCreate(objects, options = {}) {
+    const spaceId = await this._getSpaceId();
+    const objectsToCreate = objects.map(o => {
+      if (isTypeSpaceAware(o.type)) {
+        return {
+          ...o,
+          extraBodyProperties: {
+            ...o.extraBodyProperties,
+            spaceId
+          }
+        };
+      }
+      return o;
+    });
+
+    return await this._client.bulkCreate(objectsToCreate, options);
+  }
+
+  async delete(type, id) {
+    // attempt to retrieve document before deleting.
+    // this ensures that the document belongs to the current space.
+    await this.get(type, id);
+
+    return await this._client.delete(type, id);
+  }
+
+  async find(options = {}) {
+    const spaceOptions = {};
+
+    let types = options.type || this._types;
```

Comment:
Yes, I believe so: If no types are specified, then we need to filter against all known types, because we need to know which types are space-aware, and which aren't. Space-aware types need to be constrained to the current space, while non-space-aware types no not.

Reply:
Alternatively, we can say that the `find` operation will only return space-aware types, unless they explicitly request a non-space-aware type. I'd need to verify, but it might simplify the export process (I assuming we don't want to export all spaces when a user decides to export objects from a particular space)